addons/mail/controllers/poll.py

```diff
@@ -0,0 +1,112 @@
+            return
+        message = thread.message_post(
+            body="", message_type="mail_poll", subtype_xmlid="mail.mt_comment"
+        )
+        end_dt = fields.Datetime.now() + timedelta(minutes=duration)
+        poll_values = {
+            "allow_multiple_options": allow_multiple_options,
+            "option_ids": [Command.create({"option_label": label}) for label in option_labels],
+            "poll_end_dt": end_dt,
+            "poll_question": question,
+            "start_message_id": message.id,
+        }
+        # sudo - mail.poll: internal user can create poll on an accessible thread.
+        poll = self.env["mail.poll"].sudo().create(poll_values)
+        self.env.ref("mail.ir_cron_mail_end_polls")._trigger(end_dt)
+        Store(**thread._get_store_target()).add(poll).bus_send()
+        return poll.id
+
+    @route("/mail/poll/end", type="jsonrpc", auth="user", methods=["POST"])
+    def poll_end(self, poll_id):
+        poll_domain = Domain("id", "=", poll_id) & Domain("end_message_id", "=", False)
+        if not self.env.user._is_admin():
+            poll_domain &= Domain("create_uid", "=", self.env.user.id)
+        # sudo - mail.poll: creator of the poll or admins can end a poll prematurely.
+        self.env["mail.poll"].sudo().search_fetch(poll_domain)._end_and_notify()
+
+    @route("/mail/poll/delete", type="jsonrpc", auth="user", methods=["POST"])
+    def poll_delete(self, poll_id):
+        poll_domain = Domain("id", "=", poll_id)
+        if not self.env.user._is_admin():
+            poll_domain &= Domain("create_uid", "=", self.env.user.id)
+        # sudo - mail.poll: creator of the poll or admins can delete a poll.
+        self.env["mail.poll"].sudo().search_fetch(poll_domain).unlink()
+
+    @route("/mail/poll/vote", type="jsonrpc", auth="public", methods=["POST"])
+    @add_guest_to_context
+    def poll_vote(self, poll_id, option_ids):
+        options_domain = [("poll_id", "=", poll_id), ("id", "in", option_ids)]
+        # sudo - mail.poll.option: can access poll options, vote access is validated
+        # by "_get_thread_with_access_for_post" afterwards.
+        options_sudo = self.env["mail.poll.option"].sudo().search_fetch(options_domain)
+        start_message = options_sudo.poll_id.start_message_id
+        thread = self._get_thread_with_access_for_post(start_message.model, start_message.res_id)
+        if not thread:
+            return
+        user, guest = self.env["res.users"]._get_current_persona()
+        # sudo - mail.poll.vote: user can create vote on poll of accessible thread.
+        self.env["mail.poll.vote"].sudo().create(
+            [
+                {
+                    "option_id": option.id,
+                    "guest_id": guest.id,
+                    "user_id": user.id,
+                }
+                for option in options_sudo
+            ]
+        )
+        self_bus_channel = guest if self.env.user._is_public() else self.env.user
```

Comment:
user is only returned if non-public, so this can just be `user or guest` (or `guest or user` works too, as we always only return one of them)